py/redrock/rebin.py

```diff
@@ -10,6 +10,169 @@
+                double yedge = 0;
+                double r = 0;
+                double ylo, yhi;
+                //ibasis, ibin, iz = index in 3d representation of output tb array
+                int ibasis = i % nbasis;
+                int ibin = (i % (nbasis*nbin)) / nbasis;
+                int iz = i / (nbasis*nbin);
+
+                //create local vars for (1+z) and wavelength stride
+                double z = (1+myz[iz]);
+                double stride = (x[1]-x[0])*z;
+
+                //Calculate first sample beyond this bin edge
+                int j = (int)((edges[ibin]-x[0]*z)/stride)+1;
+                if (j < 0 || j >= nt) {
+                    //this should never happen but safeguard just in case
+                    result[i] = -1;
+                    return;
+                }
+
+                //Calculate the end index for this bin -- can be calculated
+                //since wavelength stride is uniform
+                int end_idx = (int)((edges[ibin+1]-x[0]*z)/stride)+1;
+                //Copy global memory to local registers and multiply by z
+                double xj = x[j]*z;
+                double xj1 = x[j-1]*z;
+                int y_idx = j+ibasis*nt;
+
+                // - What is the y value where the interpolation crossed the edge?
+                yedge = y[y_idx-1] + (edges[ibin]-xj1) * (y[y_idx]-y[y_idx-1]) / (xj-xj1);
+
+                //r = yedge;
+                // - Is this sample inside this bin?
+                if (xj < edges[ibin+1]) {
+                    area = 0.5 * (y[y_idx] + yedge) * (xj - edges[ibin]);
+                    r += area;
+
+                    //- Continue with interior bins
+                    while (j+1 < end_idx) {
+                        j++;
+                        y_idx++;
+                        xj = x[j]*z;
+                        xj1 = x[j-1]*z;
+                        area = 0.5 * (y[y_idx] + y[y_idx-1]) * (xj - xj1);
+                        r += area;
+                    }
+
+                    //- Next sample will be outside this bin; handle upper edge
+                    yedge = y[y_idx] + (edges[ibin+1]-xj) * (y[y_idx+1]-y[y_idx]) / (x[j+1]*z-xj);
+                    area = 0.5 * (yedge + y[y_idx]) * (edges[ibin+1] - xj);
+                    r += area;
+                } else {
+                    //- Otherwise the samples span over this bin
+                    ylo = y[y_idx] + (edges[ibin]-xj) * (y[y_idx] - y[y_idx-1]) / (xj - xj1);
+                    yhi = y[y_idx] + (edges[ibin+1]-xj) * (y[y_idx] - y[y_idx-1]) / (xj - xj1);
+                    area = 0.5 * (ylo+yhi) * (edges[ibin+1]-edges[ibin]);
+                    r += area;
+                }
+                result[i] = r / (edges[ibin+1]-edges[ibin]);
+            }
+
+            __global__ void batch_trapz_rebin_uneven(const double* x, const double* y, const double* edges, const double* myz, const int* idx, double* result, int nz, int nbin, int nbasis, int nt) {
```

Comment:
Evenly spaced bins is a special case of unevenly spaced bins.  Is the performance improvement from "batch_trapz_rebin" enough better than "batch_trapz_rebin_uneven" to be worth maintaining two separate kernels in addition to their python wrappers and the non-GPU numba version?